README.rst

```diff
@@ -258,44 +258,35 @@ For example:
 How To Upgrade Schema
 =====================
 
+We are using Alembic framework to facility schema migration. For more details,
+see https://alembic.sqlalchemy.org/en/latest/tutorial.html.
+
 Here's an example of a normal workflow for performing schema update.
 
-1. create a new changeset file in *pint_server/pint_db_migrate/versions/*. The
-   new changeset file must have the following format: *<d><d><d>_<string>.py*.
-   The first three digit of the filename is the version number, follow by an
-   underscore and a meaningful name of the changeset. The new changeset must
-   have the highest version number, which is usually a plus one increment of
-   the last version. For example, if *pint_server/pint_db_migrate/versions/*
-   currently contains a file *001_in_the_beginning.py*. The next changeset
-   should starts with *002*. Say if we want to add a new column *foo* to the
-   *amazonimages* table. We should create a new file named *002_add_foo_column_to_amazonimages.py* with the following content:
+1. create the Python 3.6 development virtual environment
 
    .. code-block::
 
-     from sqlalchemy import Table, MetaData, String, Column
-
-     def upgrade(migrate_engine):
-         meta = MetaData(bind=migrate_engine)
-         amazonimages = Table('amazonimages', meta, autoload=True)
-         foo = Column('foo', String(100))
-         foo.create(amazonimages)
-
-     def downgrade(migrate_engine):
-         meta = MetaData(bind=migrate_engine)
-         amazonimages = Table('amazonimages', meta, autoload=True)
-         amazonimages.c.foo.drop()
+     ./bin/create_dev_venv.sh
 
-2. create the Python 3.6 development virtual environment
+2. activate the development virtual environment
 
    .. code-block::
 
-     ./bin/create_dev_venv.sh
+     source dev_venv/bin/activate
 
-3. activate the development virtual environment
+3. create a revision
 
    .. code-block::
 
-     source dev_venv/bin/activate
+     cd public-cloud-info-service/pint_server
+     alembic revision -m 'fix some table'
+
+   If the above command is successful, you'll see the auto generate
+   revision file in `./pint_db_migrate/versions/`. The file is named
+   `<revision>_fix_some_table.py`.
+
+4. edit the revision file to do your thing
```

Comment:
This is against what was agreed upon yesterday? We don't want to be in the business of manually writing migrations.

Reply:
Not sure if I understand. There's no auto-generate SQL. Even with Alembic, only thing that is auto-generated is the framework and revisions. Schema and data migrations are very specific, I have not came across a framework that will auto-generate everything for you.

Reply:
https://alembic.sqlalchemy.org/en/latest/autogenerate.html

Reply:
Thanks for the link. Unfortunately, there are lots of caveats spell out in the doc, which IMO quick dangerous. In our case ...

1. We use Enums and PostgreSQL-specific types such as ENUM, INET, etc which cannot be auto-detected and generated.
2. It does not automatically shrink primary keys as we intended.
3. It does not do data manipulation such as removing duplicate records prior to switching primary keys.

Auto-generate in this case is more like Tesla's autopilot feature, which doesn't do exactly as it advertised, which IMO quite dangerous. :-) We can certainly add it to README to assist user as a starting point for the migration scripts. But certainly, there will be manual editing involved.

Reply:
I've never had a problem with an auto generation and I've never had to manually tweak an auto generated "schema" migration. To me this PR has a big issue in that the models still don't match the migrations.

> * We use Enums and PostgreSQL-specific types such as ENUM, INET, etc which cannot be auto-detected and generated.

This is certainly opinionated (also not directly related to the migration change) but to me this is "code smell". We are using PostgreSQL but IMHO it's best to design a data model that generically works for all (most) RDBMS.

> 
> * It does not automatically shrink primary keys as we intended.

This was caused by a communication issue regarding initial design. Ideally this is not something we would be doing in migrations. See follow on question.

> 
> * It does not do data manipulation such as removing duplicate records prior to switching primary keys.

Yes, data migrations are always manual.

Question I have is, what's the difference in effort to migrate the current DB vs starting over with a flattened migration that is generated directly from the models to ensure that we keep the two pieces in sync.

Reply:
> I've never had a problem with an auto generation and I've never had to manually tweak an auto generated "schema" migration. To me this PR has a big issue in that the models still don't match the migrations.

The schema mismatch problem we had originally was due to having ORM auto-create any missing tables by default. I've
disabled that in the patch so we would never run into schema mismatch problem again. ORM will only do data update. Any missing schema will be revealed once the ORM start inserting data.

> 
> > * We use Enums and PostgreSQL-specific types such as ENUM, INET, etc which cannot be auto-detected and generated.
> 
> This is certainly opinionated (also not directly related to the migration change) but to me this is "code smell". We are using PostgreSQL but IMHO it's best to design a data model that generically works for all (most) RDBMS.

You'll have to take it up with Robert on this one. :-) That issue was raised and got shot down. Maybe you can be more convincing than I was. :-) 

> 
> > * It does not automatically shrink primary keys as we intended.
> 
> This was caused by a communication issue regarding initial design. Ideally this is not something we would be doing in migrations. See follow on question.

I disagree. Data manipulation is not something we simply cannot avoid, especially if we are to change primary keys, unique indexes, etc again in the future, unless you want to clean up your DB manually prior to schema migration. But then that would defeat the purpose.

> 
> > * It does not do data manipulation such as removing duplicate records prior to switching primary keys.
> 
> Yes, data migrations are always manual.
> 
> Question I have is, what's the difference in effort to migrate the current DB vs starting over with a flattened migration that is generated directly from the models to ensure that we keep the two pieces in sync.

Are you proposing to nuke the production DB and start over? We may need another meeting on that. :-)

Reply:
> I've disabled that in the patch so we would never run into schema mismatch problem again

This PR initially led to mismatch. It's entirely possible, nothing is preventing it. The idea to use autogeneration was limit this possibility by generating migrations from models rather than manually creating migrations and hoping the models stay in sync.

> Data manipulation is not something we simply cannot avoid, especially if we are to change primary keys, unique indexes, etc again in the future
> Are you proposing to nuke the production DB and start over? We may need another meeting on that. :-)

Yeah, essentially that's the question. If the main concerns about using auto generation are because of the bandaids we are applying, and we can essentially start over with little effort, then those concerns go out the window.

Reply:
> This PR initially led to mismatch. It's entirely possible, nothing is preventing it. The idea to use autogeneration was limit this possibility by generating migrations from models rather than manually creating migrations and hoping the models stay in sync.
> 

I agree we should use auto-generate as a starting point. Will update README to reflect that. But sometimes we still need
edit the generated migration script if data manipulation is necessary, especially in the case where we are shrinking the primary key set.

> Yeah, essentially that's the question. If the main concerns about using auto generation are because of the bandaids we are applying, and we can essentially start over with little effort, then those concerns go out the window.

Not sure about the little effort part. That's something we'll need to work with Joel's team to come up with an estimate as this is production DB. Even if we are starting over, what are we gaining? It's still schema migration from this point onward every time we make a change. Are you saying we will never run into data problems in the future? ever? And the schema we have right now is rock solid and we'll never need to change the constraints again?

Reply:
Also, I just tried the auto generate feature, it does NOT do what we wanted. When I updated the model to remove `publishedon` from the primary keys list, the generated code does not attempt to drop the original primary key constraint, instead, it simply trying to update `nullable=True`.

```
    # ### commands auto generated by Alembic - please adjust! ###
    op.alter_column('alibabaimages', 'publishedon',
               existing_type=sa.DATE(),
               nullable=True)
    op.alter_column('alibabaimages', 'id',
               existing_type=sa.VARCHAR(length=100),
               nullable=True)
    op.alter_column('alibabaimages', 'region',
               existing_type=sa.VARCHAR(length=100),
               nullable=False)
```
This will result in errors like this when attempt to migrate.

```
Failed to upgrade Pint database: (psycopg2.errors.InvalidTableDefinition) column "publishedon" is in a primary key

[SQL: ALTER TABLE alibabaimages ALTER COLUMN publishedon DROP NOT NULL]
(Background on this error at: http://sqlalche.me/e/13/f405)
```

Bottom line, we can't trust the auto-generate code even for things such as removing a primary key.

Reply:
Right, for certain things it does not pick them up automagically. This is to be expected and is documented. For example it's tough to know if a column is renamed or if one is deleted and a different one is added. Alembic makes it clear in the generated migration:

`# ### commands auto generated by Alembic - please adjust! ###`

Just because it might miss some types of changes doesn't mean it cannot be used. Instead I think it means that the migration would be reviewed and updated accordingly if necessary.

Reply:
For the vanilla operations like adding/removing tables/columns this works fine. But anything more than that, like changing indexes, definitely required extra work. But in any case, we should use the auto-generated script as a starting point and go from there.

Reply:
Updated README.rst to always use auto generate as a starting point.